projects/RabbitMQ.Client/client/impl/Frame.cs

```diff
@@ -236,42 +251,51 @@ internal static InboundFrame ReadFrom(Stream reader)
                 ProcessProtocolHeader(reader);
             }
 
-            using (IMemoryOwner<byte> headerMemory = MemoryPool<byte>.Shared.Rent(6))
+            byte[] headerBytes = null;
+            try
             {
-                Memory<byte> headerSlice = headerMemory.Memory.Slice(0, 6);
+                headerBytes = ArrayPool<byte>.Shared.Rent(6);
```

Comment:
stackalloc?

Reply:
@stebet I'm assuming that instead of using `new byte[6]` as a field in the class you're doing this because it uses the memory buffer pool (reducing allocation churn) and supports `.Slice(2)` on this line:

```
int payloadSize = NetworkOrderDeserializer.ReadInt32(headerSlice.Slice(2));
```

Is that a correct understanding?

Reply:
Yes. stackalloc might work there. Good catch. Also, @davidfowl , I'd there a simple way to reuse/cache a TaskCompletionSource? I'm thinking of the ConcurrentQueue work-item notifier in the AsyncEventingSubscriber. SemaphoreSlim seemed to perform a lot worse for signaling than just completing a TCS and creating a new one.

Reply:
Added support for stackalloc.

Reply:
ConcurrentQueue + SemaphoreSlim = Channels. It’ll do what you need and be more efficient.

Reply:
Adding channels in https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/835